pkg/cmd/listen.go

```diff
@@ -96,181 +101,149 @@ Stripe account.`,
-					Connect:        false,
-					EventTypes:     endpoint.EnabledEvents,
-				})
-			} else {
-				endpointRoutes = append(endpointRoutes, proxy.EndpointRoute{
-					URL:            buildForwardURL(forwardConnectURL, u),
-					ForwardHeaders: forwardConnectHeaders,
-					Connect:        true,
-					EventTypes:     endpoint.EnabledEvents,
-				})
-			}
+	for el := range proxyOutCh {
+		err := el.Accept(proxyVisitor)
+		if err != nil {
+			return err
 		}
 	}
 
-	return endpointRoutes
+	return nil
 }
 
-// parseURL parses the potentially incomplete URL provided in the configuration
-// and returns a full URL
-func parseURL(url string) string {
-	_, err := strconv.Atoi(url)
-	if err == nil {
-		// If the input is just a number, assume it's a port number
-		url = "localhost:" + url
-	}
-
-	if strings.HasPrefix(url, "/") {
-		// If the input starts with a /, assume it's a relative path
-		url = "localhost" + url
-	}
+func withSIGTERMCancel(ctx context.Context, onCancel func()) context.Context {
+	// Create a context that will be canceled when Ctrl+C is pressed
+	ctx, cancel := context.WithCancel(ctx)
 
-	if !strings.HasPrefix(url, "http://") && !strings.HasPrefix(url, "https://") {
-		// Add the protocol if it's not already there
-		url = "http://" + url
-	}
+	interruptCh := make(chan os.Signal, 1)
+	signal.Notify(interruptCh, os.Interrupt, syscall.SIGTERM)
 
-	return url
+	go func() {
+		<-interruptCh
+		onCancel()
+		cancel()
+	}()
+	return ctx
 }
 
-func buildForwardURL(forwardURL string, destination *url.URL) string {
-	f, err := url.Parse(forwardURL)
-	if err != nil {
-		log.Fatalf("Provided forward url cannot be parsed: %s", forwardURL)
-	}
+func createVisitor(logger *log.Logger, format string, printJSON bool) *websocket.Visitor {
+	var s *spinner.Spinner
+
+	return &websocket.Visitor{
+		VisitError: func(ee websocket.ErrorElement) error {
+			ansi.StopSpinner(s, "", logger.Out)
+			logger.Fatal(ee.Error)
+			return ee.Error
+		},
+		VisitStatus: func(se websocket.StateElement) error {
+			switch se.State {
+			case websocket.Loading:
+				s = ansi.StartNewSpinner("Getting ready...", logger.Out)
+			case websocket.Reconnecting:
+				ansi.StartSpinner(s, "Session expired, reconnecting...", logger.Out)
+			case websocket.Ready:
+				ansi.StopSpinner(s, fmt.Sprintf("Ready! Your webhook signing secret is %s (^C to quit)", ansi.Bold(se.Data[0])), logger.Out)
+			case websocket.Done:
+				ansi.StopSpinner(s, "", logger.Out)
+			}
+			return nil
+		},
+		VisitData: func(de websocket.DataElement) error {
+			stripeEvent, ok := de.Data.(proxy.StripeEvent)
+			if !ok {
+				return fmt.Errorf("VisitData received unexpected type for DataElement, got %T expected %T", de, proxy.StripeEvent{})
+			}
 
-	return fmt.Sprintf(
-		"%s://%s%s%s",
-		f.Scheme,
-		f.Host,
-		strings.TrimSuffix(f.Path, "/"), // avoids having a double "//"
-		destination.Path,
-	)
+			if strings.ToUpper(format) == outputFormatJSON || printJSON {
+				fmt.Println(de.Marshaled)
```

Comment:
Do we want this to print the line or go to the logger?

Reply:
this should probably go to the logger yes, will fix!

Reply:
actually, this is the actual events being printed, and using the logger changes the output (timestamp, log level, etc).
We know that some users are probably parsing those logs (especially with json format) and changing the format will probably be a backward incompatible change. Do we want to?

Reply:
cc @tomer-stripe I'd like to hear what you think? I'm not sure what's the implication of using the logger vs fmt here, as it's all getting printed to the terminal from the cobra command for the user. All this logic has mostly been moved, so it's probably safer to keep `fmt` where it's been used?

Reply:
So I'm thinking more broadly for the gRPC work -- we probably wouldn't want that to print anything if they're running in gRPC mode right? (since all the output should go to the gRPC stream itself). A log-level output would be helpful for debugging though!

It's probably a larger scope ask but I think what the CLI actually needs is a separate package that handles all output and it's just automatically done based on how it's being called (gRPC, CLI, debug, etc).

Reply:
Probably makes sense at some point to do some additional cleanup - I think eventually what we want is
- core logic only _returns_ values/errors
- client (cobra, gRPC, etc) processes returns and handles it however it wants (prints, log, etc).

That being said, this section is actually a `cobra` visitor, so it's only ran when listen is ran through the CLI, so we always do want to print here. This loc will never be called when listen is ran through gRPC.